Add ability to abort NCCL communicators from the store.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32895 Add ability to abort NCCL communicators from the store.**

When a particular rank calls `ncclCommAbort` on a communicator, it is
important to ensure all other ranks call `ncclCommAbort` on their respective
communicators. If this is not done, the other ranks could get stuck causing the
GPU to spin with 100% utilization.

To alleviate this issue, whenever any rank calls `ncclCommAbort` we put the
unique communicator id in the store. The NCCL watchdog thread then monitors the
store and aborts any communicators found in the store as "aborted".

A few more general fixes in this PR:

1) Use std::shared_ptr for the store in PrefixStore. PrefixStore was using a
reference to the store and when that reference went out of scope the store
object it was holding onto was invalid. This caused a segfault in the watchdog
thread.
2) Enhanced logging for the watchdog thread.

Differential Revision: [D19638159](https://our.internmc.facebook.com/intern/diff/D19638159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19638159/)!

Commit:
Add ability to abort NCCL communicators from the store.

When a particular rank calls `ncclCommAbort` on a communicator, it is
important to ensure all other ranks call `ncclCommAbort` on their respective
communicators. If this is not done, the other ranks could get stuck causing the
GPU to spin with 100% utilization.

To alleviate this issue, whenever any rank calls `ncclCommAbort` we put the
unique communicator id in the store. The NCCL watchdog thread then monitors the
store and aborts any communicators found in the store as "aborted".

A few more general fixes in this PR:

1) Use std::shared_ptr for the store in PrefixStore. PrefixStore was using a
reference to the store and when that reference went out of scope the store
object it was holding onto was invalid. This caused a segfault in the watchdog
thread.
2) Enhanced logging for the watchdog thread.

Differential Revision: [D19638159](https://our.internmc.facebook.com/intern/diff/D19638159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19638159/)!

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 29cc0771:


**None of the build failures appear to be your fault**.

* **2/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=29cc0771cbcea1d1fff344ce38fc5e31f0e19399).

### :snowflake: 2 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4395813) pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94015959) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94015959)) :snowflake:
<details>
<summary>
<code>Feb 01 03:04:17 RuntimeError: Process 0 terminated or timed out after 200.2006425857544 seconds</code>
</summary>

```
Feb 01 03:04:17 ====================================================================== 
Feb 01 03:04:17 ERROR [200.209s]: test_nccl_timeout (__main__.NcclErrorHandlingTest) 
Feb 01 03:04:17 ---------------------------------------------------------------------- 
Feb 01 03:04:17 Traceback (most recent call last): 
Feb 01 03:04:17   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 174, in wrapper 
Feb 01 03:04:17     self._join_processes(fn) 
Feb 01 03:04:17   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 255, in _join_processes 
Feb 01 03:04:17     self._check_return_codes(elapsed_time) 
Feb 01 03:04:17   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 274, in _check_return_codes 
Feb 01 03:04:17     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time)) 
Feb 01 03:04:17 RuntimeError: Process 0 terminated or timed out after 200.2006425857544 seconds 
Feb 01 03:04:17  
Feb 01 03:04:17 ---------------------------------------------------------------------- 
Feb 01 03:04:17 Ran 117 tests in 1512.577s 
Feb 01 03:04:17  
Feb 01 03:04:17 FAILED (errors=7, skipped=7) 
Feb 01 03:04:17  
Feb 01 03:04:17 Generating XML reports... 
Feb 01 03:04:17 Traceback (most recent call last): 
Feb 01 03:04:17   File "test/run_test.py", line 457, in <module> 
Feb 01 03:04:17     main() 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4395811) pytorch_linux_xenial_cuda10_1_cudnn7_py3_NO_AVX2_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94016100) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94016100)) :snowflake:
<details>
<summary>
<code>Feb 01 03:05:00 RuntimeError: Process 0 terminated or timed out after 200.2006332874298 seconds</code>
</summary>

```
Feb 01 03:05:00 ====================================================================== 
Feb 01 03:05:00 ERROR [200.208s]: test_nccl_timeout (__main__.NcclErrorHandlingTest) 
Feb 01 03:05:00 ---------------------------------------------------------------------- 
Feb 01 03:05:00 Traceback (most recent call last): 
Feb 01 03:05:00   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 174, in wrapper 
Feb 01 03:05:00     self._join_processes(fn) 
Feb 01 03:05:00   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 255, in _join_processes 
Feb 01 03:05:00     self._check_return_codes(elapsed_time) 
Feb 01 03:05:00   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 274, in _check_return_codes 
Feb 01 03:05:00     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time)) 
Feb 01 03:05:00 RuntimeError: Process 0 terminated or timed out after 200.2006332874298 seconds 
Feb 01 03:05:00  
Feb 01 03:05:00 ---------------------------------------------------------------------- 
Feb 01 03:05:00 Ran 117 tests in 1511.359s 
Feb 01 03:05:00  
Feb 01 03:05:00 FAILED (errors=7, skipped=7) 
Feb 01 03:05:00  
Feb 01 03:05:00 Generating XML reports... 
Feb 01 03:05:01 Traceback (most recent call last): 
Feb 01 03:05:01   File "test/run_test.py", line 457, in <module> 
Feb 01 03:05:01     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "Add ability to abort NCCL communicators from the store."


When a particular rank calls `ncclCommAbort` on a communicator, it is
important to ensure all other ranks call `ncclCommAbort` on their respective
communicators. If this is not done, the other ranks could get stuck causing the
GPU to spin with 100% utilization.

To alleviate this issue, whenever any rank calls `ncclCommAbort` we put the
unique communicator id in the store. The NCCL watchdog thread then monitors the
store and aborts any communicators found in the store as "aborted".

A few more general fixes in this PR:

1) Use std::shared_ptr for the store in PrefixStore. PrefixStore was using a
reference to the store and when that reference went out of scope the store
object it was holding onto was invalid. This caused a segfault in the watchdog
thread.
2) Enhanced logging for the watchdog thread.

Differential Revision: [D19638159](https://our.internmc.facebook.com/intern/diff/D19638159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19638159/)!

[ghstack-poisoned]

zhaojuanmao commented:
I'm wondering why this solution works while aborting timed out work from watchdog does not work? both of them aborted ncclCommunicators and removed them from the cache, right?

pritamdamania87 commented:
The earlier solution didn't work since the watchdog would erase work from the `oustandingWork_` queue and that would trigger `~WorkNCCL` which in turn would trigger some CUDA cleanup operations resulting in the destructor getting stuck and as a result the watchdog thread got stuck too. Note that in this solution we don't remove the communicators from the cache. The reason is that if we do that, other ranks might have failed, but one rank might call allreduce which would try to initialize a new communicator (since old ones were deleted) and this operation would get stuck forever.

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -300,14 +340,56 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
+          // wait more than the watchdog sleep time.).
+          const auto& storeKey = getNcclAbortedCommStoreKey(commId);
+          try {
+            store_->wait(
```
What's the default interval you're running this at?

There are 2 caveats with using the store at runtime:
1) If you have a large number of processes all hitting the store at a frequent interval this may cause problems.
2) If you use the TCP store on rank 0, this can timeout/fail if that's the process that went down.

Otherwise, this seems like a decent fix here.

Could you add a couple paragraphs explaining why out of band messaging for aborting NCCL comms is needed? This will help out future readers (including perhaps yourself) on why all of this is needed.

pietern reviewed (commented):
Leaving approval to @mrshenli @zhaojuanmao.

zhaojuanmao commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -212,6 +229,10 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
         // can not run new events successfully.
         for (const auto& ncclComm : ncclComms_) {
           ncclComm->ncclCommAbort();
```
here we abort ncclComm as well, but we did not add it to abortedComms_() because this will trigger nccl error and will be caught by watchdog, then watch dog will add this id to abortedComms_, right?

zhaojuanmao commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -300,14 +340,56 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
+    }
 
-          // Remove communicators from the cache.
-          it = devNCCLCommMap_.erase(it);
```
would you please also explain why we do not need to erase the cache? IIUC, since all processes will abort ncclComm, application codes should be able to catch exception and destroy process group?

zhaojuanmao reviewed (approved):
approving to unblock, please also follow up @pietern's comments as well, my inline comments are mostly for clarification

pritamdamania committed:
Update on "Add ability to abort NCCL communicators from the store."


When a particular rank calls `ncclCommAbort` on a communicator, it is
important to ensure all other ranks call `ncclCommAbort` on their respective
communicators. If this is not done, the other ranks could get stuck causing the
GPU to spin with 100% utilization.

To alleviate this issue, whenever any rank calls `ncclCommAbort` we put the
unique communicator id in the store. The NCCL watchdog thread then monitors the
store and aborts any communicators found in the store as "aborted".

A few more general fixes in this PR:

1) Use std::shared_ptr for the store in PrefixStore. PrefixStore was using a
reference to the store and when that reference went out of scope the store
object it was holding onto was invalid. This caused a segfault in the watchdog
thread.
2) Enhanced logging for the watchdog thread.

Differential Revision: [D19638159](https://our.internmc.facebook.com/intern/diff/D19638159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19638159/)!

[ghstack-poisoned]
